snapfaas/src/worker.rs

```diff
@@ -27,6 +27,93 @@ pub struct Worker {
+    let mut i = 0;
+    let result = loop {
+        let mut tsps = tsps.clone();
+        if i == 5 {
+            break RequestStatus::ProcessRequestFailed;
+        }
+        i += 1;
+        let (tx, rx) = mpsc::channel();
+        vm_req_sender.send(Message::GetVm(function_name.clone(), tx)).expect("Failed to send GetVm request");
+        match rx.recv().expect("Failed to receive GetVm response") {
+            Ok(mut vm) => {
+                tsps.allocated = precise_time_ns();
+                if !vm.is_launched() {
+                    // newly allocated VM is returned, launch it first
+                    if let Err(e) = vm.launch(Some(func_req_sender.clone()), vm_listener.try_clone().expect("clone unix listener"), cid, false, None) {
+                        handle_vm_error(e);
+                        let _ = rsp_sender.send(Response {
+                            status: RequestStatus::LaunchFailed,
+                        });
+                        // a VM launched or not occupies system resources, we need
+                        // to put back the resources assigned to this VM.
+                        vm_req_sender.send(Message::DeleteVm(vm)).expect("Failed to send DeleteVm request");
+                        // insert the request's timestamps
+                        stat.push(tsps);
+                        continue;
+                    }
+                }
+
+                debug!("VM is launched");
+                tsps.launched = precise_time_ns();
+
+                match vm.process_req(req.payload.clone()) {
+                    Ok(rsp) => {
+                        tsps.completed = precise_time_ns();
+                        // TODO: output are currently ignored
+                        debug!("{:?}", rsp);
+                        vm_req_sender.send(Message::ReleaseVm(vm)).expect("Failed to send ReleaseVm request");
+                        break RequestStatus::SentToVM(rsp);
+                    }
+                    Err(e) => {
+                        handle_vm_error(e);
+                        vm_req_sender.send(Message::DeleteVm(vm)).expect("Failed to send DeleteVm request");
+                        // insert the request's timestamps
+                        stat.push(tsps);
+                        continue;
+                    },
+                }
+
+            },
+            Err(e) => {
+                let id = thread::current().id();
+                let status = match e {
+                    resource_manager::Error::InsufficientEvict |
+                    resource_manager::Error::LowMemory(_) => {
+                        error!("[Worker {:?}] Resource exhaustion", id);
+                        RequestStatus::ResourceExhausted
+                    }
+                    resource_manager::Error::FunctionNotExist=> {
+                        error!("[Worker {:?}] Requested function doesn't exist: {:?}", id, function_name);
+                        RequestStatus::FunctionNotExist
```

Comment:
A minor optimization: we can break out of loop if the error is FunctionNotExist. Retrying cannot solve it.